signer/signer.go

```diff
@@ -0,0 +1,134 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	msgsigner "github.com/form3tech-oss/go-http-message-signatures"
+	"github.com/form3tech-oss/http-message-signing-proxy/config"
+	"github.com/form3tech-oss/http-message-signing-proxy/proxy"
+)
+
+const (
+	requestTargetHeaderKey = "(request-target)"
+	digestHeaderKey        = "digest"
+)
+
+type DataError = msgsigner.DataError
+
+type requestSigner struct {
+	*msgsigner.MessageSigner
+	headerConfig config.HeadersConfig
+}
+
+func NewRequestSigner(cfg config.SignerConfig) (proxy.RequestSigner, error) {
+	key, err := loadKey(cfg.KeyFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	signatureHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	digestHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := msgsigner.NewRSASigner(key, signatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	msgSigner, err := msgsigner.NewMessageSigner(digestHashAlgo, signer, cfg.KeyId, msgsigner.Authorization)
+	if err != nil {
+		return nil, err
+	}
+
+	return &requestSigner{
+		MessageSigner: msgSigner,
+		headerConfig:  cfg.Headers,
+	}, err
+}
+
+func (rs *requestSigner) SignRequest(req *http.Request) (*http.Request, error) {
+	headers, err := rs.getSignatureHeaders(req)
+	if err != nil {
+		return nil, err
+	}
+
+	signedReq, err := rs.MessageSigner.SignRequest(req, headers)
+	switch err.(type) {
+	case *msgsigner.DataError, *msgsigner.SigningError:
+		return nil, proxy.NewInvalidRequestError(err)
+	default:
+		return signedReq, err
+	}
+}
+
+func (rs *requestSigner) getSignatureHeaders(req *http.Request) ([]string, error) {
+	// Get the intersection of request's headers and config's signature headers
+	var headers []string
+	for _, header := range rs.headerConfig.SignatureHeaders {
+		if req.Header.Get(header) != "" {
```

Comment:
```suggestion
		if req.Header.Get(http.CanonicalHeaderKey(header)) != "" {
```

Reply:
In `Header.Get`, it calls `textproto.MIMEHeader(h).Get(key)` which is basically the same thing?

Reply:
That's just a type, it doesn't make the `header` the correct case https://pkg.go.dev/net/textproto#MIMEHeader

Reply:
Ah, just read the docs about it being case insensitive, cool :)

Reply:
I think it does 
```
func (h MIMEHeader) Get(key string) string {
	if h == nil {
		return ""
	}
	v := h[CanonicalMIMEHeaderKey(key)]
	if len(v) == 0 {
		return ""
	}
	return v[0]
}
```